src/bindings/js/node/src/core_wrap.cpp

```diff
@@ -126,28 +141,37 @@ Napi::Value CoreWrap::compile_model_sync(const Napi::CallbackInfo& info,
 }
 
 Napi::Value CoreWrap::compile_model_sync_dispatch(const Napi::CallbackInfo& info) {
+    std::vector<std::string> errors_messages;
+
+    // Allowed signatures list
+    static auto path_and_device = NapiArg::Validator().add_string_arg().add_string_arg();
+    static auto path_device_and_config = NapiArg::Validator().add_string_arg().add_string_arg().add_object_arg();
+    static auto model_and_device = NapiArg::Validator().add_object_arg().add_string_arg();
+    static auto model_device_and_config = NapiArg::Validator().add_object_arg().add_string_arg().add_object_arg();
+
     try {
-        if (info.Length() == 2 && info[0].IsString() && info[1].IsString()) {
+        if (path_and_device.validate(info, errors_messages)) {
             return compile_model_sync(info, info[0].ToString(), info[1].ToString());
-        } else if (info.Length() == 2 && info[0].IsObject() && info[1].IsString()) {
+        } else if (model_and_device.validate(info, errors_messages)) {
             return compile_model_sync(info, info[0].ToObject(), info[1].ToString());
```

Comment:
My proposal would be to use template based approach with unpacking of arguments to provide simplified and clear interface.

Let's start with `args_validate<...>` which will be also universal for whole JS API. Here `CustomValue` and vector are replacing actual napi's `info`. **Also returning `0`/throwing errors can be replaced with `True/False`/error_messages, this is just a showcase:**
```cpp
#include <iostream>
#include <vector>
#include <stdexcept>

class CustomValue {
public:
    CustomValue(double value) : m_value(value) {}

    CustomValue(std::string value) : m_string(value) {}

    int to_int() {
        return static_cast<int>(m_value);
    }

    std::string to_string() {
        if (!m_string.empty()) {
            return m_string;   
        }
        return std::to_string(m_value);
    }

private:
    double m_value;
    std::string m_string;
};

template<typename T>
void args_validate_impl(CustomValue arg) {
    throw std::runtime_error("Validation for this type is not implemented!");
}

template<>
void args_validate_impl<std::string>(CustomValue arg) {
    // You can replace it with check_type here:
    auto arg_str = arg.to_string();
    if (arg_str.compare(std::string("Jano")) != 0) {
        throw std::runtime_error("This is not me!");
    }
}

template<>
void args_validate_impl<int>(CustomValue arg) {
    // You can replace it with check_type here:
    auto arg_int = arg.to_int();
    if (arg_int != 6) {
        throw std::runtime_error("I do not like this number!");
    }
}

// Forward declaration of the recursive case
template<typename T, typename... Args>
int args_validate_detail(std::vector<CustomValue>& info, int index);

// Base case for a single type and a single argument
template<typename T>
int args_validate_single(std::vector<CustomValue>& info, int index) {
    args_validate_impl<T>(info[index]);
    return 0;
}

// Recursive case for multiple types and arguments
template<typename T, typename... Args>
int args_validate_detail(std::vector<CustomValue>& info, int index) {
    args_validate_impl<T>(info[index]);
    // Call the next function in the chain
    return args_validate_single<Args...>(info, index + 1);
}

// Specialization of the recursive case -- only one type left
template<typename T>
int args_validate_detail(std::vector<CustomValue>& info, int index) {
    // Call the base case function
    return args_validate_single<T>(info, index);
}

template<typename T, typename... Ts>
int args_validate(std::vector<CustomValue>& info) {
    return args_validate_detail<T, Ts...>(info, 0);
}

// Call to it
int main() {
    std::vector<CustomValue> info{CustomValue("Jano"), CustomValue(6.0)};
    try {
        // Try different values inside `info` or change types in template
        auto c = args_validate<std::string, int>(info);
        std::cout << c << '\n';
    } catch (const std::runtime_error& e) {
        std::cerr << e.what() << '\n';
    }
    return 0;
}
```

That would result in something like this, mixing either C++/OV/custom JS or just go full on "js args":
```cpp
if (args_validate<std::string, std::string>(info)) {
    return compile_model_sync<std::string, std::string>(info);
}
else if (args_validate<ov::Model, js::arg::device>(info)) {
    return compile_model_sync<ov::Model, js::arg::device>(info);
}
```

What is the difference?
- First you remove responsibility to write allowed signature list over and over again for each function. The responsibility is moved to if-else tree which will most likely stay in every scenario.
- Reduce the points of errors, now only if-else tree has the responsibility to dispatch data.
- `compile_model_sync` has clear call to it, just unpack and feed it into the function, no need to remember which `X_and_Y` is `info[0].ToA, info[1].ToB`. (here templates must be implemented, probably in similar fashion -- that further improvement which will align with above approach)

Personally I think it will be easier to maintain one `common_validation.cpp/hpp` file with all overloads, given that they will be reusable across whole API. What are you thoughts on that?

Reply:
@vishniakov-nikolai 

> Why do we need to replace Napi::CallbackInfo by some other class?

There is no need. My code is just functional example that can be easily compiled even online, that's for reference and general idea explanation.

> Current implementation in the case of incorrect arguments gives errors like this.
> Is it possible to get similar error messages from proposed solution?

I am pretty sure that can be added as well. Is there anything stopping from passing down the error-storage? It will virtually work the same way.

> With recursive parameters checking you will not repeat some code, but have to review all of checks, one by one diving into the depths.

I am not sure about this part, could you explain it in more details? As far as I understand your code, you are also going into helper functions like `model_and_device` and also pass around it if fails to match, so all arguments must match in both approaches (mine and yours).

> could you implement your proposal based on example from js-binding code?

That's a question to @mlukasze @almilosz and you -- should I spent time on this real case implementation or delegate it? @almilosz are you up for the task? I can consult "on the fly".